File: process.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strings"
	"sync"
	// "bytes"
	"io"
)

const (
	chunksize int = 10
)

func main() {
	sourceDir, err := filepath.Abs(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	destDir, err := filepath.Abs(os.Args[2])
	if err != nil {
		log.Fatal(err)
	}

	if _, err := os.Stat(destDir); os.IsNotExist(err) {
		if err := os.Mkdir(destDir, 0777); err != nil {
			log.Fatal(err)
		}
	}

	files, err := ioutil.ReadDir(sourceDir)
	if err != nil {
		log.Fatal(err)
	}

	var wg sync.WaitGroup

	countFileLines := func(fileName, sourceDir, destDir string) {
		defer wg.Done()
		data, err := os.Open(filepath.Join(sourceDir, fileName))
		if err != nil {
			log.Fatal(err)
		}

		defer data.Close()
		var sentence string

		reader := bufio.NewReader(data)
		buffer := make([]byte, chunksize)
		var count int
		for {
			if count, err = reader.Read(buffer); err != nil {
				break
			}
			sentence += string(buffer[:count])
		}

		if err != io.EOF {
			log.Fatal("Error Reading ", fileName, ": ", err)
		} else {
			err = nil
		}

		newFileName := strings.TrimSuffix(fileName, filepath.Ext(fileName)) + ".res"
		f, err := os.Create(filepath.Join(destDir, newFileName))

		if err != nil {
			log.Fatal("create file ", err)
		}

		defer f.Close()
		if _, err = f.WriteString(sentence); err != nil {
			log.Fatal(err)
		}
	}

	for _, file := range files {
		wg.Add(1)
		go countFileLines(file.Name(), sourceDir, destDir)
	}

	wg.Wait()
	fmt.Println("Total number of processed files:", len(files))
}

File: process.js
'use strict'

const fs = require('fs')
const path = require('path');

function countFileLines (fileName, sourceDir, destDir) {
    const filePath = `${sourceDir}/${fileName}`
    return new Promise((resolve, reject) => {
    let lineCount = 0;
    fs.createReadStream(filePath)
      .on("data", (buffer) => {
        let idx = -1;
        do {
          idx = buffer.indexOf(10, idx+1);
          lineCount++;
        } while (idx !== -1);
      }).on("end", () => {
        const extension = path.extname(fileName);
        const newFileName = path.basename(fileName,extension);
        const pathToOutputFile = `${destDir}/${newFileName}.res`;
        fs.writeFile(pathToOutputFile, lineCount, (err) => {
          if (err) reject(err);
          resolve();
        });
        resolve(lineCount);
      }).on("error", reject);
    });
  };


(async function main(){
    let pathToSourceDir;
    let pathToDestinationDir;
    try {
        pathToSourceDir = path.resolve(process.argv[2]);
      } catch (err) {
        console.error('Wrong source directory path!');
        process.exit(1);
      }
    
      try {
        pathToDestinationDir = path.resolve(process.argv[3]);
      } catch (err) {
        console.error('Wrong destination directory path!');
      }
      if(!fs.existsSync(path.resolve(process.argv[3]))) {
          console.warn('Create dir\t' +process.argv[3])
          fs.mkdirSync(path.resolve(process.argv[3]))
        }
         
    const inputFiles = fs.readdirSync(pathToSourceDir);
    const promise = [];
    for (const file of inputFiles) {
        const res = countFileLines(file, pathToSourceDir, pathToDestinationDir);
        promise.push(res)
    }
    await Promise.all(promise);
})();